File: String & Array/3Sum Closest/solution.ts
function threeSumClosest(nums: number[], target: number): number {
    let result = 0;
    let distance = Number.MAX_SAFE_INTEGER;
    nums.sort((a,b) => a-b);
    for(let i = 0; i < nums.length; i++) {
        const picked = nums[i];
        const remaining = target - nums[i];
        let first = i+1;
        let last = nums.length - 1;
        while(first < last) {

            const total = nums[first] + nums[last];
            if(total - remaining === 0) return picked + total ;
            if(Math.abs(total - remaining) < distance) {
                distance = Math.abs(total - remaining);
                result = picked + total;
            }
            if(total < remaining) first++;
            if(total > remaining) last--;
        }
            
    }
    return result;
};

File: String & Array/3Sum Closest/solution.go
package main

import (
	"math"
	"sort"
)

func threeSumClosest(nums []int, target int) int {
	min := 999999
	result := 0
	sort.Ints(nums)
	for i := 0; i < len(nums)-2; i++ {
		firstNum := nums[i]
		newTarget := target - firstNum
		secondIndex := i + 1
		thirdIndex := len(nums) - 1
		for secondIndex < thirdIndex {
			total := nums[secondIndex] + nums[thirdIndex]
			distance := int(math.Abs(float64(newTarget - total)))
			if distance < min {
				result = total + firstNum
				min = distance
				if min == 0 {
					return total + firstNum
				}
			}
			if total < newTarget {
				secondIndex++
			} else {
				thirdIndex--
			}
		}
	}
	return result
}
